extensions/librdkafka/PublishKafka.h

```diff
@@ -193,121 +186,146 @@ class PublishKafka : public core::Processor {
-                 rd_kafka_topic_t *rkt,
-                 rd_kafka_t *rk,
-                 const std::shared_ptr<core::FlowFile> flowFile,
+    ReadCallback(const uint64_t max_seg_size,
+                 std::string key,
+                 rd_kafka_topic_t * const rkt,
+                 rd_kafka_t * const rk,
+                 std::shared_ptr<core::FlowFile> flowFile,
                  utils::Regex &attributeNameRegex,
                  std::shared_ptr<Messages> messages,
-                 size_t flow_file_index)
-        : max_seg_size_(max_seg_size),
-          key_(key),
+                 const size_t flow_file_index,
+                 const bool fail_empty_flow_files)
+        : flowFile_(std::move(flowFile)),
+          flow_size_(flowFile_->getSize()),
+          max_seg_size_(max_seg_size == 0 || flow_size_ < max_seg_size ? flow_size_ : max_seg_size),
+          key_(std::move(key)),
           rkt_(rkt),
           rk_(rk),
-          flowFile_(flowFile),
+          hdrs(make_headers()),
           messages_(std::move(messages)),
           flow_file_index_(flow_file_index),
-          attributeNameRegex_(attributeNameRegex) {
-      flow_size_ = flowFile_->getSize();
-      status_ = 0;
-      read_size_ = 0;
-      hdrs = nullptr;
-    }
+          status_(0),
+          read_size_(0),
+          attributeNameRegex_(attributeNameRegex),
+          fail_empty_flow_files_(fail_empty_flow_files)
+    { }
 
     ~ReadCallback() {
-      if (hdrs) {
-        rd_kafka_headers_destroy(hdrs);
-      }
+      rd_kafka_headers_destroy(hdrs);
     }
 
-    int64_t process(std::shared_ptr<io::BaseStream> stream) {
-      if (max_seg_size_ == 0U || flow_size_ < max_seg_size_) {
-        max_seg_size_ = flow_size_;
-      }
+    int64_t process(const std::shared_ptr<io::BaseStream> stream) {
       std::vector<unsigned char> buffer;
-      buffer.reserve(max_seg_size_);
+
+      buffer.resize(max_seg_size_);
       read_size_ = 0;
       status_ = 0;
-      rd_kafka_resp_err_t err;
+      called_ = true;
 
-      for (auto kv : flowFile_->getAttributes()) {
-        if(attributeNameRegex_.match(kv.first)) {
-          if (!hdrs) {
-            hdrs = rd_kafka_headers_new(8);
-          }
-          err = rd_kafka_header_add(hdrs, kv.first.c_str(), kv.first.size(), kv.second.c_str(), kv.second.size());
-        }
+      assert(max_seg_size_ != 0 || flow_size_ == 0 && "max_seg_size_ == 0 implies flow_size_ == 0");
+      // ^^ therefore checking max_seg_size_ == 0 handles both division by zero and flow_size_ == 0 cases
+      const size_t reserved_msg_capacity = max_seg_size_ == 0 ? 1 : utils::intdiv_ceil(flow_size_, max_seg_size_);
+      messages_->modifyResult(flow_file_index_, [reserved_msg_capacity](FlowFileResult& flow_file) {
+        flow_file.messages.reserve(reserved_msg_capacity);
+      });
+
+      // If the flow file is empty, we still want to send the message, unless the user wants to fail_empty_flow_files_
```

Comment:
The default value is the property is now true, which means in this case we drop the files. 
This is not a backward compatible behaviour.

Reply:
It is backwards compatible, but not desired. That behavior is why the issue exists.

Reply:
see these 3 lines for proof: https://github.com/apache/nifi-minifi-cpp/pull/712/files#diff-9917deef91cc1503aff82b472310b7eaL662

Reply:
Ah, okay, thanks!